### PR TITLE
Add websocket support

### PR DESCRIPTION
**Pull request by annaloukianova**
tangentially related to #33

### CODE REVIEW

File: src/websocket.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      });
+
+      emitter.on('abort', () => socket.close());
+    } catch (err) {
+      emitter.emit('error', err);
+    }
+  });
+
+  return emitter;
+};
```
Comment (helfi92) [inline src/websocket.js]: This is really nice. Thoughts about the following diff + the comment from above:

```diff
diff --git a/src/websocket.js b/src/websocket.js
index d6079a2..c7dcc83 100644
--- a/src/websocket.js
+++ b/src/websocket.js
@@ -13,31 +13,39 @@ export default (url, options) => {
 
   emitter.on('start', () => {
     try {
+      const { onOpen, onClose, formatMessage } = options;
+
       // try to connect to websocket
       const socket = new WebSocket(url);
 
+      // websocket event 1
       socket.addEventListener('open', event => {
-        if (options.onOpen) {
-          options.onOpen(event, socket);
+        if (onOpen) {
+          onOpen(event, socket);
         }
       });
-      socket.addEventListener('close', options.onClose);
-
+      // websocket event 2
+      socket.addEventListener('close', event => {
+        if (onClose) {
+          onClose(event);
+        }
+      });
+      // websocket event 3
       socket.addEventListener('error', error => {
         emitter.emit('error', error);
       });
-
+      // websocket event 4
       socket.addEventListener('message', event => {
         let message = event.data;
 
-        if (options.formatMessage) {
-          message = options.formatMessage(message);
+        if (formatMessage) {
+          message = formatMessage(message);
         }
 
         emitter.emit('data', message);
       });
 
-      emitter.on('abort', () => socket.close());
+      emitter.on('abort', socket.close);
     } catch (err) {
       emitter.emit('error', err);
     }

```

File: src/components/LazyLog/index.jsx [hunk 2]
```diff
@@ -44,11 +45,22 @@ export default class LazyLog extends Component {
      * Defaults to `{ credentials: 'omit' }`.
      */
     fetchOptions: object,
+    /**
+     * Options object which will be passed through to websocket.
+     * Available options are: `onOpen: (event, socket)`, `onClose: (event)`,
+     * `onError: (event)`, `formatMessage: (message: event.data)`.
+     */
+    websocketOptions: object,
```
Comment (helfi92) [inline src/components/LazyLog/index.jsx]: We can make this a bit more explicit by replacing `object` with `shape`:

```js
websocketOptions: shape({
  // add comment about onOpen
  onOpen: func,
  // add comment about onClose
  onClose: func,
  // add comment about onError
  onError: func,
  // add comment about formatMessage
  formatMessage: func,
});
```

File: src/websocket.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      socket.addEventListener('open', event => {
+        if (options.onOpen) {
+          options.onOpen(event, socket);
+        }
+      });
+      socket.addEventListener('close', options.onClose);
+
+      socket.addEventListener('error', error => {
+        emitter.emit('error', error);
+      });
```
Comment (helfi92) [inline src/websocket.js]: I think the error emitter should only be called when there is an error loading the actual url which should get caught in the catch statement.

https://github.com/mozilla-frontend-infra/react-lazylog/blob/de294c475d622221234542bdbc38ffde3f9c9255/src/components/LazyLog/index.jsx#L101-L105

I guess we can just call the onError handler in websocketOptions if it exists. Something like:

```js
const { onError } = options;

// ...

socket.addEventListener('error', error => {
  if (onError) {
    onError(error);
  }
});
```

**Comment by helfi92** [comment]
@annaloukianova Did you want to complete the PR by any chance? :-)

**Comment by brendanjryan** [comment]
Is this PR still under active development? I am very interested in this feature and would love to pick up the work here as a collaborator or submit another patch if possible!

**Comment by helfi92** [comment]
> Is this PR still under active development? I am very interested in this feature and would love to pick up the work here as a collaborator or submit another patch if possible!

Go for it. This PR probably just needs a rebase and the requested changes from the review. Let me know if you had any questions. Thanks!

**Pull request closed without merging**